node/src/manager/commands/chain.rs

```diff
@@ -143,3 +152,55 @@ pub fn remove(primary: ConnectionPool, store: Arc<BlockStore>, name: String) ->
 
     Ok(())
 }
+
+pub fn change_block_cache_shard(
+    primary_store: ConnectionPool,
+    store: Arc<BlockStore>,
+    chain_name: String,
+    shard: String,
+) -> Result<(), Error> {
+    println!("Changing block cache shard for {} to {}", chain_name, shard);
+
+    let conn = primary_store.get()?;
+    let chain = find_chain(&conn, &chain_name)?.expect("no such chain");
+    let old_shard = chain.shard;
+
+    println!("Current shard: {}", old_shard);
+
+    let chain_store = store
+        .chain_store(&chain_name)
+        .ok_or_else(|| anyhow!("unknown chain: {}", chain_name))?;
+    let new_name = format!("{}-old", &chain_name);
+
+    conn.transaction(|| -> Result<(), StoreError> {
+        let ident = chain_store.chain_identifier.clone();
+        let shard = Shard::new(shard.to_string())?;
+        // Drop the foriegn key constraint on deployment_schemas
+        sql_query(
+            "alter table deployment_schemas drop constraint deployment_schemas_network_fkey;",
+        )
+        .execute(&conn)?;
+
+        // Update the current chain name to chain-old
+        update_chain_name(&conn, &chain_name, &new_name)?;
+        chain_store.update_name(&new_name)?;
+
+        // Create a new chain with the name in the destination shard
+        let chain=  add_chain(&conn, &chain_name, &ident, &shard)?;
+        store.add_chain_store(&chain, ChainStatus::Ingestible, true)?;
+
+        // Re-add the foreign key constraint
+        sql_query(
+            "alter table deployment_schemas add constraint deployment_schemas_network_fkey foreign key (network) references chains(name);",
+        )
+         .execute(&conn)?;
+        Ok(())
+    })?;
```

Comment:
One thing I am wondering about is what happens when this block gets interrupted (e.g., because the user presses ^C) Since there are three shards involved, it may leave everything in a messy state, possibly in a state where `graph-node` won't start again. I think it's right to use one transaction for the primary, at least that way, we won't lose the constraint.

Maybe this order of operations is safer:

1. Add the new chain store to the new shard (no change to primary yet) It's a little sucky since you can't use the `add_chain` helper for that as that modifies the primary
2. Update the `chains` table in the primary (in one txn: update `chains` entry to `-old`, create new entry for chain name pointing to new shard)
3. Update the name in the old shard to `-old`

I am not 100% sure that that leaves everything at least in a state where `graph-node` will start again, but I think it's close.

Reply:
Im not entirely sure of what you suggested, i've made some changes now, could you take a look.

Changed the order, but had to get the next value for the chains_id_seq when doing this before calling `add_chain`

https://github.com/graphprotocol/graph-node/blob/8cded389543e972753baafa258814aa60d7f478e/node/src/manager/commands/chain.rs#L183-L221

Reply:
What I am trying to get at is that the `conn.transaction` block puts changes in the primary into a transaction. If that gets interrupted, the transaction in the primary gets rolled back, but changes made in other shards will stick around. My concern then is that we could end up in a state where the primary still points at the chain in the old shard (in `chains`), but we've already updated the chain name in that shard (in `ethereum_networks`), and so when `graph-node` starts up again, it won't find that chain since it first looks at the `chains` table to figure out which shard the block cache is in, but can't find the corresponding entry in `ethereum_networks`.

I _think_ the only change that's needed from what you have right now is to move the call to `chain_store.update_name(&new_name)?;` out of the transaction block. When the txn commits we have an entry for `chain` in the primary that points to the new shard and is correct. We'd also have an entry for `chain-old` that points at the old shard and is incorrect (`ethereum_networks` there still has the entry for `chain`), but that's tolerable since `graph-node` _should_ still start up though I am not 100% confident about that.

Unfortunately, this is really tricky, and testing all possible points where an interrupt (or error, e.g., because the db crashed) could happen is very hard.